common/src/main/java/org/apache/celeborn/common/network/sasl/CelebornSaslClient.java

```diff
@@ -0,0 +1,164 @@
+import java.util.Map;
+
+import javax.annotation.Nullable;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.RealmCallback;
+import javax.security.sasl.RealmChoiceCallback;
+import javax.security.sasl.Sasl;
+import javax.security.sasl.SaslClient;
+import javax.security.sasl.SaslException;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A SASL Client for Celeborn which simply keeps track of the state of a single SASL session, from
+ * the initial state to the "authenticated" state. This client initializes the protocol via a
+ * firstToken, which is then followed by a set of challenges and responses.
+ *
+ * <p>TODO: Currently, we have hardcoded Sasl mechanism to be DIGEST-MD5 for connection auth. We
+ * should make this configurable in the future (see:
+ * https://issues.apache.org/jira/browse/CELEBORN-1158).
+ */
+public class CelebornSaslClient {
+  private static final Logger logger = LoggerFactory.getLogger(CelebornSaslClient.class);
+
+  private SaslClient saslClient;
+
+  public CelebornSaslClient(
+      String saslMechanism,
+      @Nullable Map<String, String> saslProps,
+      @Nullable CallbackHandler authCallbackHandler) {
+    Preconditions.checkNotNull(saslMechanism);
+    try {
+      this.saslClient =
+          Sasl.createSaslClient(
+              new String[] {saslMechanism},
+              null,
+              null,
+              DEFAULT_REALM,
+              saslProps,
+              authCallbackHandler);
+    } catch (SaslException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** Used to initiate SASL handshake with server. */
+  public synchronized byte[] firstToken() {
+    if (saslClient != null && saslClient.hasInitialResponse()) {
+      try {
+        return saslClient.evaluateChallenge(new byte[0]);
+      } catch (SaslException e) {
+        throw new RuntimeException(e);
+      }
+    } else {
+      return new byte[0];
+    }
+  }
+
+  /** Determines whether the authentication exchange has completed. */
+  public synchronized boolean isComplete() {
+    return saslClient != null && saslClient.isComplete();
+  }
+
+  /** Returns the value of a negotiated property. */
+  public Object getNegotiatedProperty(String name) {
+    return saslClient.getNegotiatedProperty(name);
+  }
+
+  /**
+   * Respond to server's SASL token.
+   *
+   * @param token contains server's SASL token
+   * @return client's response SASL token
+   */
+  public synchronized byte[] response(byte[] token) {
+    try {
+      return saslClient != null ? saslClient.evaluateChallenge(token) : new byte[0];
+    } catch (SaslException e) {
+      throw Throwables.propagate(e);
```

Comment:
nit `Throwables.propagate` method has been marked as deprecated in higher versions of guava.

Reply:
done.